### ISSUE TITLE
Memory leaking

### ISSUE THREAD

**Issue by jrquirk** [alcap-org/AlcapDAQ#152]
I think the memory leaks we have might be transient because our goal is still to use a navigator. But I ran

```
valgrind --leak-check=full ./rootana -i ~edmonds_a/data/tree/tree02164.root -o out02164.root -m production.cfg
```

On our [v1.2](https://github.com/alcap-org/AlcapDAQ/blob/rootana_v1.2/analyzer/rootana/production.cfg) production modules file. There were lots of thing, but the biggest flags were sent up by

```
==14091== 1,014,938,480 bytes in 12,686,731 blocks are definitely lost in loss record 48,637 of 48,638
==14091==    at 0x4A075BC: operator new(unsigned long) (vg_replace_malloc.c:298)
==14091==    by 0x514DA88: TStorage::ObjectAlloc(unsigned long) (TStorage.cxx:332)
==14091==    by 0x42B34F: MaxBinAPGenerator::ProcessPulses(std::vector<TPulseIsland*, std::allocator<TPulseIsland*> > const&, std::vector<TAnalysedPulse*, std::allocator<TAnalysedPulse*> >&) (TObject.h:156)
==14091==    by 0x426451: MakeAnalysedPulses::ProcessEntry(TGlobalData*, TSetupData*) (MakeAnalysedPulses.cpp:132)
==14091==    by 0x445C7E: BaseModule::ProcessGenericEntry(TGlobalData*, TSetupData*) (BaseModule.cpp:54)
==14091==    by 0x41C671: Main_event_loop(TTree*, ARGUMENTS&) (main.cpp:228)
==14091==    by 0x41CE0A: main (main.cpp:153)
==14091== 
==14091== 1,014,941,200 bytes in 12,686,765 blocks are definitely lost in loss record 48,638 of 48,638
==14091==    at 0x4A075BC: operator new(unsigned long) (vg_replace_malloc.c:298)
==14091==    by 0x514DA88: TStorage::ObjectAlloc(unsigned long) (TStorage.cxx:332)
==14091==    by 0x42E8A9: SimpIntAPGenerator::ProcessPulses(std::vector<TPulseIsland*, std::allocator<TPulseIsland*> > const&, std::vector<TAnalysedPulse*, std::allocator<TAnalysedPulse*> >&) (TObject.h:156)
==14091==    by 0x426451: MakeAnalysedPulses::ProcessEntry(TGlobalData*, TSetupData*) (MakeAnalysedPulses.cpp:132)
==14091==    by 0x445C7E: BaseModule::ProcessGenericEntry(TGlobalData*, TSetupData*) (BaseModule.cpp:54)
==14091==    by 0x41C671: Main_event_loop(TTree*, ARGUMENTS&) (main.cpp:228)
==14091==    by 0x41CE0A: main (main.cpp:153)
==14091== 
==14091== LEAK SUMMARY:
==14091==    definitely lost: 2,151,918,848 bytes in 26,898,972 blocks
==14091==    indirectly lost: 4,592 bytes in 160 blocks
==14091==      possibly lost: 187,905 bytes in 5,142 blocks
==14091==    still reachable: 8,717,897 bytes in 81,577 blocks
==14091==         suppressed: 0 bytes in 0 blocks
==14091== Reachable blocks (those to which a pointer was found) are not shown.
==14091== To see them, rerun with: --leak-check=full --show-reachable=yes
==14091== 
==14091== For counts of detected and suppressed errors, rerun with: -v
==14091== ERROR SUMMARY: 380 errors from 292 contexts (suppressed: 7916 from 18)
```

I just can't figure out how to follow this line

```
==14091==    by 0x42B34F: MaxBinAPGenerator::ProcessPulses(std::vector<TPulseIsland*, std::allocator<TPulseIsland*> > const&, std::vector<TAnalysedPulse*, std::allocator<TAnalysedPulse*> >&) (TObject.h:156)
```

My guess would be that the TAPs aren't getting deleted like they should, but that could be off base. In the [main event loop](https://github.com/alcap-org/AlcapDAQ/blob/rootana_v1.2/analyzer/rootana/src/main.cpp#L214-L219) perhaps the wrong clear method is getting called? Because the [ClearGlobalData](https://github.com/alcap-org/AlcapDAQ/blob/rootana_v1.2/analyzer/rootana/src/main.cpp#L279-L290) function looks like it ostensibly cleans this up.

Specifically I was hoping maybe @litchfld or @benkrikler might be able to glean more from this? Or if this seems right?

I'll try the obvious of rerunning to test this, but was hoping to catch one of you guys before you hit the old dusty trail.

**Comment by litchfld** [comment]
It looks like only the TPulseIslands are getting wiped.  ClearGlobalData was disabled in a commit with this message: 

> The god-forsaken ClearGlobalData method was giving me grief again, so I'm commenting it out of main completely.  If we hit memory leaks we can revisit this.

:laughing: So it's a feature not a bug.  I'm inclined to agree with (past) Ben, leave it for now, at least until we want to do a new rootana production. Only a few files have failed, and it's better to come round to it systematically than layer patch on patch.

**Comment by litchfld** [comment]
As for this line:

```
==14091==    by 0x42B34F: 
MaxBinAPGenerator::ProcessPulses(vector<TPulseIsland*> const&, vector<TAnalysedPulse*>&)
 (TObject.h:156)
```

Its a little hard to follow because there are some hidden indirections in this, but i think what is happening:
- Starting at [line 32](https://github.com/alcap-org/AlcapDAQ/blob/develop/analyzer/rootana/src/TAP_generators/MaxBinAPGenerator.cpp#L32) of MaxBinAPGenerator, this function called is [inlined](https://github.com/alcap-org/AlcapDAQ/blob/develop/analyzer/rootana/src/TAP_generators/MaxBinAPGenerator.cpp#L32-L37)  in TVAnalysedPulse So there's probably no machine code corresponding to the call.  
- The function [_template_](https://github.com/alcap-org/AlcapDAQ/blob/develop/analyzer/rootana/src/TAP_generators/TVAnalysedPulseGenerator.h#L72) this wraps is not strictly code - the real code is generated by the compliler from this [template fragment](https://github.com/alcap-org/AlcapDAQ/blob/develop/analyzer/rootana/src/TAP_generators/TVAnalysedPulseGenerator.h#L119-127), so there is technically no source to refer to here either.
- ~~In there I guess the pointer `pulse` is not created (optimiser puts it straight at the return site), and i guess the temporary pointer `parent` is similarly elided. So the first thing that is actually done is (the template instantiation of) `pulse=new TypeOfTAP(GetSource(),parent_index,parent);`, where `TypeOfTAP` is replaced by `TAnalysedPulse`~~  
  - **EDIT**: Actually I don't think this matters and may be wrong. The point is that the stack frame isn't available in source, not that nothing else happens  
- Here we have a `new TAnalysedPulse`, which inherits from TObject.  Base obejects are instantiated first, and excitingly TObject overloads [`operator new`](http://root.cern.ch/root/html/src/TObject.h.html#156) at line 156 of TObject.h.

So valgrind is saying "the bit of memory allocated at TObject.h:L156" is the source of the leak (we can assume everything after that is okay, otherwise ROOT would be shot through with leaks), and pointing to the nearest bit of code that calls it. It just happens the nearest thing it can point to is a long way back.

**Comment by litchfld** [comment]
One thing i'm a little uncertain of though. I would have assumed that the real payload of TAnalysedPulse would involve more memory than needed for the TObject base.  So either i'm wrong and there's another (smaller) leak for the rest of the TAP, or that part of the object is getting cleaned up through some seriously funky deletion.

**Comment by litchfld** [comment]
PS Thanks for this John, I was going to have a peek today but that (a) saved me a lot of slog, and (b) left the fun bit still to do. :wink:

**Comment by jrquirk** [comment]
I assume this was taken care of in the new navigator, and so am closing this issue.

**Comment by litchfld** [comment]
Not as far as I know... as I said in the pull request #157, atomising the output is still to do.

**Comment by jrquirk** [comment]
I've run rootana through valgrind several times since, and the memory leak seems to have been taken care of.
